packages/trie/test/proof/rightNode.spec.ts

```diff
@@ -0,0 +1,83 @@
+import { utf8ToBytes } from 'ethereum-cryptography/utils'
+import { describe, expect, it } from 'vitest'
+
+import { BranchNode, Trie } from '../../src/index.js'
+import { hasRightBranch, isValueNode, returnRightNode } from '../../src/proof/range.js'
+import { bytesToNibbles, nibblesCompare, nibblestoBytes } from '../../src/util/nibbles.js'
+describe('returns RightNode', async () => {
+  const trie = new Trie()
+  const entries = Array.from({ length: 20 }, (_, i) => {
+    return {
+      key: utf8ToBytes(`key${i % 3}${i % 4}${i % 5}`),
+      value: Uint8Array.from([i, i, i, i]),
+    }
+  })
+  for (const { key, value } of entries) {
+    await trie.put(key, value)
+  }
+
+  const sortedByKey = entries.sort((a, b) =>
+    nibblesCompare(bytesToNibbles(a.key), bytesToNibbles(b.key))
+  )
+  it('passes', () => {
+    expect(true).toBe(true)
+  })
+  for (const { key, value } of sortedByKey) {
+    it(`returns RightNode for key ${key} (${value})`, async () => {
+      const val = await trie.get(key)
+      const node = (await trie.findPath(key)).node
+      expect(val).toEqual(value)
+      expect(isValueNode(node!)).toBe(true)
+    })
+  }
+})
+
+describe('hasRightBranch', async () => {
+  const branchNode = new BranchNode()
+  for (const i of Array.from({ length: 7 }, (_, i) => i)) {
+    branchNode.setBranch(i, Uint8Array.from([i]))
+  }
+  branchNode.setBranch(8, Uint8Array.from([8]))
+  for (const i of Array.from({ length: 16 }, (_, i) => i)) {
+    if (i < 8) {
+      it(`hasRightBranch for ${i}`, async () => {
+        expect(hasRightBranch(branchNode, i)).toBeDefined()
+      })
+    } else {
+      it(`hasRightBranch for ${i}`, async () => {
+        expect(hasRightBranch(branchNode, i)).toBeFalsy()
+      })
+    }
+  }
+})
+describe('returnRightNode', async () => {
+  const trie = new Trie({})
+  for (const i of Array.from({ length: 9 }, (_, i) => i)) {
+    if (i === 6) continue
+    await trie.put(nibblestoBytes([0, i, i, i]), Uint8Array.from([i]))
+  }
+  for (const i of Array.from({ length: 10 }, (_, idx) => idx)) {
+    if (i > 7) {
+      it(`returnRightNode (null) for ${i}`, async () => {
+        const rightNode = await returnRightNode.bind(trie)(
+          bytesToNibbles(nibblestoBytes([0, i, i, i]))
+        )
+        try {
+          expect(rightNode).toBeNull()
+        } catch {
+          expect(rightNode!.node).toBeUndefined()
+        }
+      })
+    } else {
+      const rightNode = await returnRightNode.bind(trie)(
+        bytesToNibbles(nibblestoBytes([0, i, i, i]))
+      )
+      it(`returnRightNode for ${i} = ${i === 5 ? i + 2 : i + 1}`, async () => {
+        expect(rightNode, ``).not.toBeNull()
+        i === 5
+          ? expect(rightNode?.node.value()).toEqual(Uint8Array.from([i + 2]))
+          : expect(rightNode?.node.value()).toEqual(Uint8Array.from([i + 1]))
```

Comment:
I have no idea what's going on here. How does the trie look? Why are there conditionals here for these values? Why is index 5 special? I don't see this in the trie setup.

Reply:
*
                                          BRANCH_NODE
    0     1     2     3     4     5     6     7    8    9    10    11   12   13   14   15
    *     *     *      *     *     *     _     *     _    _      _      _      _      _     _      _
    ^     ^     ^      ^     ^     ^            ^
                       LEAF_NODES
    
    
    We have a branch node, with leaves on branches 0, 1, 2, 3, 4, 5, and 7.  This tests the basic ability of `returnRightNode` to work from different starting conditions:
    
  -  Keys 0 - 4 :   There is a node in the Trie at that address, and the next node to the right is one index higher.
  -  Key 5:  There is a node in the Trie at that address, and a node to its right, but with an empty branch between them.  This tests that `rightNode` isn't simply looking 1 index higher, but will continue searching to find a value node.
  - Key 6:  There is NOT a node in the Trie at that address, but there is a node to its right.  This tests that `rightNode` can start its search correctly with or without finding a node for that key.
  - Key 7, there is a node at the address, but 0 nodes to the right.
  - Key 8 - 15, there is no node at the address, and 0 nodes to the right.